File: example/xx_three/lines_fat.cpp
#ifdef TWO_MODULES
module;
#include <infra/Cpp20.h>
#include <xx_three/ex.h>
module two.xxthree;
#else
#include <xx_three/xx_three.h>
#include <gfx-pbr/Api.h>
#include <stl/array.h>
#include <stl/vector.hpp>
#include <cstdio>
#include <cstring>
#endif

using namespace two;

#define FAT 1
#define CACHE 0

EX(xx_lines_fat)
{
#if UI
	UNUSED(dockbar);
	SceneViewer& viewer = ui::scene_viewer(parent);
	Scene& scene = viewer.m_scene;
#else
	static Scene scene = Scene(app.m_gfx);
	static GfxViewer viewer = GfxViewer(window, scene);
#endif

#if UI
	ui::orbit_controls(viewer);
	//controls.minDistance = 10;
	//controls.maxDistance = 500;
#endif

	//camera2 = new THREE.PerspectiveCamera(40, 1, 1, 1000);
	//camera2.position.copy(camera.position);
	
	static Material* material = nullptr;

#if !CACHE
	static Lines lines;
	static Batch* batch = nullptr;
#endif

	if(init)
	{
		Camera& camera = viewer.m_camera;
		camera.m_fov = 40.f; camera.m_near = 1.f; camera.m_far = 1000.f;
		camera.m_eye = vec3(-40.f, 0.f, 60.f);

		auto setup_material = [&](Material& m) {
			m.m_solid.m_colour = rgb(0xffffff);
			m.m_line.m_line_width = 5.f;
			m.m_line.m_dashed = true;
			m.m_base.m_shader_color = ShaderColor::Vertex;
		};

		const vector<vec3> points = hilbert3d(vec3(0.f), 20.f, 1);

		const CurveCatmullRom3 curve = { points };
		const uint32_t divisions = uint32_t(12U * points.size());

		const float l = float(divisions);

#if CACHE
		Lines lines;
#endif
		lines.start(curve.point(0.f), hsl(0.f, 1.f, 0.5f));

		for(uint32_t i = 0; i < divisions; i++)
		{
			const vec3 position = curve.point(float(i) / l);
			const Colour color = hsl(float(i) / l, 1.f, 0.5f);

			lines.next(position, color);
		}

#if FAT
			// fat line

			Program& program = app.m_gfx.programs().fetch("line_fat");
			material = &app.m_gfx.materials().create("line_fat", setup_material);
			material->m_program = &program;

			const Model& model = app.m_gfx.models().fetch("line");

			Node3& n = gfx::nodes(scene).add(Node3());
			Item& it = gfx::items(scene).add(Item(n, model, 0U, material));
#if !CACHE
			batch = &gfx::batches(scene).add(Batch(it, sizeof(Lines::Segment)));
			it.m_batch = batch;
#else
			Batch& batch = gfx::batches(scene).add(Batch(it, sizeof(Lines::Segment)));
			it.m_batch = &batch;
#endif

#if CACHE
			batch.cache({ (float*)lines.m_segments.data(), lines.m_segments.size() * sizeof(Lines::Segment) / sizeof(float) });
#endif
#else
			// regular line

			Program& program = app.m_gfx.programs().fetch("line");
			material = &app.m_gfx.materials().create("line", setup_material);
			material->m_program = &program;
			
			Model& model = app.m_gfx.create_model("lines");
			lines.write(*model.m_items[0].m_mesh);

			Node3& n = gfx::nodes(scene).add(Node3());
			Item& it = gfx::items(scene).add(Item(n, model, 0U, material));
#endif

	}

#if !CACHE
	// this should not be necessary each frame, add cache to the batch ?
	lines.commit(*batch);
#endif

#if UI
	if(Widget* dock = ui::dockitem(dockbar, "Game", { 1U }))
	{
		Widget& sheet = ui::columns(*dock, { 0.3f, 0.7f });

		ui::slider_field(sheet, "line width", material->m_line.m_line_width, { 0.f, 20.f, 0.1f });

		ui::field<bool>(sheet, "dashed", material->m_line.m_dashed);
		ui::slider_field(sheet, "dash scale", material->m_line.m_dash_scale, { 0.f, 20.f, 0.1f });
		ui::slider_field(sheet, "dash size",  material->m_line.m_dash_size,  { 0.f, 20.f, 0.1f });
		ui::slider_field(sheet, "dash gap",   material->m_line.m_dash_gap,   { 0.f, 20.f, 0.1f });
	}
#endif
}
			

File: example/xx_three/two/lines_fat.js
// lines_fat.js

var viewer = two.ui.scene_viewer(panel);
two.ui.orbit_controller(viewer);
//controls.minDistance = 10;
//controls.maxDistance = 500;

var scene = viewer.scene;

function hilbert3d(center, size, iterations, v0, v1, v2, v3, v4, v5, v6, v7) {

    // Default Vars
    var center = center !== undefined ? center : { x: 0, y: 0, z: 0 },
        size = size !== undefined ? size : 10,
        half = size / 2,
        iterations = iterations !== undefined ? iterations : 1,
        v0 = v0 !== undefined ? v0 : 0,
        v1 = v1 !== undefined ? v1 : 1,
        v2 = v2 !== undefined ? v2 : 2,
        v3 = v3 !== undefined ? v3 : 3,
        v4 = v4 !== undefined ? v4 : 4,
        v5 = v5 !== undefined ? v5 : 5,
        v6 = v6 !== undefined ? v6 : 6,
        v7 = v7 !== undefined ? v7 : 7
    ;
    
    var px = center.x + half; var py = center.y + half; var pz = center.z + half;
    var nx = center.x - half; var ny = center.y - half; var nz = center.z - half;
    var corners = [
        { x: nx, y: py, z: nz },
        { x: nx, y: py, z: pz },
        { x: nx, y: ny, z: pz },
        { x: nx, y: ny, z: nz },
        { x: px, y: ny, z: nz },
        { x: px, y: ny, z: pz },
        { x: px, y: py, z: pz },
        { x: px, y: py, z: nz }
    ];

    var vec = [ corners[v0], corners[v1], corners[v2], corners[v3], corners[v4], corners[v5], corners[v6], corners[v7] ];

    // Recurse iterations
    if (--iterations >= 0) {

        var tmp = [];

        Array.prototype.push.apply(tmp, hilbert3d(vec[0], half, iterations, v0, v3, v4, v7, v6, v5, v2, v1));
        Array.prototype.push.apply(tmp, hilbert3d(vec[1], half, iterations, v0, v7, v6, v1, v2, v5, v4, v3));
        Array.prototype.push.apply(tmp, hilbert3d(vec[2], half, iterations, v0, v7, v6, v1, v2, v5, v4, v3));
        Array.prototype.push.apply(tmp, hilbert3d(vec[3], half, iterations, v2, v3, v0, v1, v6, v7, v4, v5));
        Array.prototype.push.apply(tmp, hilbert3d(vec[4], half, iterations, v2, v3, v0, v1, v6, v7, v4, v5));
        Array.prototype.push.apply(tmp, hilbert3d(vec[5], half, iterations, v4, v3, v2, v5, v6, v1, v0, v7));
        Array.prototype.push.apply(tmp, hilbert3d(vec[6], half, iterations, v4, v3, v2, v5, v6, v1, v0, v7));
        Array.prototype.push.apply(tmp, hilbert3d(vec[7], half, iterations, v6, v5, v2, v1, v0, v3, v4, v7));

        // Return recursive call
        return tmp;
    }

    // Return complete Hilbert Curve.
    return vec;
}

if (init) {
    var camera = viewer.camera;
    camera.fov = 40.0; camera.near = 1.0; camera.far = 1000.0;
    camera.eye = new two.vec3(-40.0, 0.0, 60.0);

    var program = app.gfx.programs.fetch('line');

    var material = app.gfx.materials.create('line'); var m = material;
        m.program = program;
        m.solid.colour = two.rgb(0xffffff);
        m.line.line_width = 5.0;
        m.line.dashed = true;
        m.base.shader_color = two.ShaderColor.Vertex;

    this.lines = new two.Lines(app.gfx);

    var points = hilbert3d(new two.vec3(0.0), 20.0, 1);
    var curve = new two.CurveCatmullRom3(points);

    for(var i = 0; i < points.length; i++)
    {
        var p = points[i];
        curve.add_point(new two.vec3(p.x, p.y, p.z));
    }

    var div = Math.round(12 * points.length);
    
    this.lines.start(curve.point(0.0), two.hsl(0.0, 1.0, 0.5));
    
    for(var i = 0; i < div; i++)
    {
        var p = curve.point(i / div);
        var c = two.hsl(i / div, 1.0, 0.5);
        
        this.lines.next(p, c);
    }

    var n = scene.nodes().add(new two.Node3());
    var it = scene.items().add(new two.Item(n, this.lines.model, 0, material));
    this.batch = scene.batches().add(new two.Batch(it));
    it.batch = this.batch;
}

var root = scene.begin();
two.gfx.radiance(root, 'radiance/tiber_1_1k.hdr', two.BackgroundMode.Radiance);

this.lines.commit(this.batch);
